File: src/domain/methodology.ts
export type IsoDateTime = string;
export type ConfidenceLevel = "low" | "medium" | "high";
export type DraftReviewSeverity = "low" | "medium" | "high";
export type ChangeType = "create" | "update" | "remove";

export interface SourceReference {
  sourceId: string;
  title: string;
  locator: string;
}

export interface EvidenceLocator {
  locator: string;
  summary: string;
}

export interface EvidenceCoverage {
  scope: string;
  notes: string;
}

export interface ConfidenceRating {
  level: ConfidenceLevel;
  rationale: string;
}

export interface EvidenceCard {
  id: string;
  kind: "evidence-card";
  source: SourceReference;
  evidence: EvidenceLocator[];
  observedTechnique: string;
  sourceFunction: string;
  transferableMechanism: string;
  forbiddenCopy: string[];
  coverage: EvidenceCoverage;
  confidence: ConfidenceRating;
  createdAt: IsoDateTime;
  updatedAt: IsoDateTime;
}

export interface ReaderContract {
  id: string;
  kind: "reader-contract";
  promise: string;
  progress: string;
  payoff: string;
  tone: string;
  emotionalTarget: string;
  evidenceCardIds: string[];
}

export interface KnowledgeBoundary {
  id: string;
  kind: "knowledge-boundary";
  pov: string;
  narratorKnows: string[];
  readerKnows: string[];
  withheldInformation: string[];
  unreliable: boolean;
}

export interface SettingPressure {
  id: string;
  kind: "setting-pressure";
  settingElement: string;
  restriction: string;
  resource: string;
  cost: string;
  choiceImpact: string;
}

export interface ReferenceMechanism {
  id: string;
  kind: "reference-mechanism";
  evidenceCardId: string;
  mechanism: string;
  transferRules: string[];
  forbiddenCopies: string[];
}

export interface SceneCard {
  id: string;
  kind: "scene-card";
  workId: string;
  title: string;
  pov: string;
  knowledgeBoundaryId: string;
  settingPressureIds: string[];
  readerContractId: string;
  characterGoals: string[];
  conflictAgenda: string;
  revealedInformation: string[];
  hiddenInformation: string[];
  stateChange: string;
  turn: string;
}

export interface DraftReviewItem {
  id: string;
  kind: "draft-review-item";
  targetId: string;
  severity: DraftReviewSeverity;
  category: string;
  finding: string;
  recommendation: string;
  evidenceCardIds: string[];
}

export interface ChangeSetEntry {
  targetId: string;
  changeType: ChangeType;
  description: string;
}

export interface ChangeSet {
  id: string;
  kind: "change-set";
  createdAt: IsoDateTime;
  summary: string;
  changes: ChangeSetEntry[];
  affectedIds: string[];
}

export interface ContextPack {
  id: string;
  kind: "context-pack";
  createdAt: IsoDateTime;
  updatedAt: IsoDateTime;
  purpose: string;
  evidenceCards: EvidenceCard[];
  readerContracts: ReaderContract[];
  knowledgeBoundaries: KnowledgeBoundary[];
  settingPressures: SettingPressure[];
  referenceMechanisms: ReferenceMechanism[];
  sceneCards: SceneCard[];
  draftReviewItems: DraftReviewItem[];
  changeSets: ChangeSet[];
}

export interface EvidenceCardValidationResult {
  valid: boolean;
  missing: string[];
}

const REQUIRED_EVIDENCE_CARD_FIELDS = [
  "evidence",
  "source",
  "coverage",
  "confidence",
] as const;

export function validateEvidenceCard(
  value: Partial<EvidenceCard>,
): EvidenceCardValidationResult {
  const missing: string[] = REQUIRED_EVIDENCE_CARD_FIELDS.filter(
    (field) => value[field] == null,
  );

  if (value.evidence != null && value.evidence.length === 0) {
    missing.push("evidence");
  }
  if (value.evidence != null) {
    value.evidence.forEach((evidence, index) => {
      if (!hasText(evidence.locator)) {
        missing.push(`evidence[${index}].locator`);
      }
      if (!hasText(evidence.summary)) {
        missing.push(`evidence[${index}].summary`);
      }
    });
  }
  if (value.source != null) {
    if (!hasText(value.source.sourceId)) {
      missing.push("source.sourceId");
    }
    if (!hasText(value.source.title)) {
      missing.push("source.title");
    }
    if (!hasText(value.source.locator)) {
      missing.push("source.locator");
    }
  }
  if (value.coverage != null) {
    if (!hasText(value.coverage.scope)) {
      missing.push("coverage.scope");
    }
    if (!hasText(value.coverage.notes)) {
      missing.push("coverage.notes");
    }
  }
  if (value.confidence != null) {
    if (value.confidence.level == null) {
      missing.push("confidence.level");
    }
    if (!hasText(value.confidence.rationale)) {
      missing.push("confidence.rationale");
    }
  }

  return {
    valid: missing.length === 0,
    missing: [...missing],
  };
}

function hasText(value: string | undefined): boolean {
  return typeof value === "string" && value.length > 0;
}

export function createToyContextPack(timestamp: IsoDateTime): ContextPack {
  const evidenceCard: EvidenceCard = {
    id: "evidence-toy-pressure",
    kind: "evidence-card",
    source: {
      sourceId: "source-public-toy",
      title: "Public Toy Source",
      locator: "chapter-1-scene-1",
    },
    evidence: [
      {
        locator: "chapter-1-scene-1",
        summary:
          "Invented public example: a gate opens only after a limited token is spent.",
      },
    ],
    observedTechnique: "A visible resource limit turns a simple route into a choice.",
    sourceFunction:
      "Creates pressure without needing copied events, names, or dialogue.",
    transferableMechanism:
      "Expose a scarce resource before the character commits to a route.",
    forbiddenCopy: [
      "Do not copy source events, names, dialogue, or relationship chains.",
    ],
    coverage: {
      scope: "single-scene",
      notes: "Covers only the pressure mechanism in a toy scene.",
    },
    confidence: {
      level: "medium",
      rationale: "The example is narrow and intentionally public-safe.",
    },
    createdAt: timestamp,
    updatedAt: timestamp,
  };

  return {
    id: "context-pack-toy",
    kind: "context-pack",
    createdAt: timestamp,
    updatedAt: timestamp,
    purpose: "Toy original planning context",
    evidenceCards: [evidenceCard],
    readerContracts: [
      {
        id: "reader-contract-toy",
        kind: "reader-contract",
        promise: "A practical problem will reveal character priorities.",
        progress: "Each choice shows a cost.",
        payoff: "The final route resolves the immediate constraint.",
        tone: "focused",
        emotionalTarget: "curiosity",
        evidenceCardIds: [evidenceCard.id],
      },
    ],
    knowledgeBoundaries: [
      {
        id: "knowledge-boundary-toy",
        kind: "knowledge-boundary",
        pov: "third-person-limited",
        narratorKnows: ["The current room condition."],
        readerKnows: ["A token is required."],
        withheldInformation: ["The later consequence of spending the token."],
        unreliable: false,
      },
    ],
    settingPressures: [
      {
        id: "setting-pressure-toy",
        kind: "setting-pressure",
        settingElement: "Token gate",
        restriction: "Only one path can be opened.",
        resource: "One-use token",
        cost: "The alternate path closes.",
        choiceImpact: "The protagonist must choose which clue to preserve.",
      },
    ],
    referenceMechanisms: [
      {
        id: "reference-mechanism-toy",
        kind: "reference-mechanism",
        evidenceCardId: evidenceCard.id,
        mechanism: "Scarcity before commitment",
        transferRules: ["Keep the scarcity abstract.", "Invent new setting details."],
        forbiddenCopies: ["Names", "Dialogue", "Event sequence"],
      },
    ],
    sceneCards: [
      {
        id: "scene-card-toy",
        kind: "scene-card",
        workId: "work-public-toy",
        title: "Toy Gate Choice",
        pov: "third-person-limited",
        knowledgeBoundaryId: "knowledge-boundary-toy",
        settingPressureIds: ["setting-pressure-toy"],
        readerContractId: "reader-contract-toy",
        characterGoals: ["Preserve the most useful clue."],
        conflictAgenda: "Resource scarcity forces prioritization.",
        revealedInformation: ["The gate consumes the token."],
        hiddenInformation: ["The closed path still matters later."],
        stateChange: "The protagonist commits to one path.",
        turn: "A tool becomes a cost.",
      },
    ],
    draftReviewItems: [
      {
        id: "draft-review-toy",
        kind: "draft-review-item",
        targetId: "scene-card-toy",
        severity: "medium",
        category: "continuity",
        finding: "Clarify that the token is consumed.",
        recommendation: "Add one original sentence before the gate opens.",
        evidenceCardIds: [evidenceCard.id],
      },
    ],
    changeSets: [
      {
        id: "change-set-toy",
        kind: "change-set",
        createdAt: timestamp,
        summary: "Clarified the toy gate cost.",
        changes: [
          {
            targetId: "scene-card-toy",
            changeType: "update",
            description: "Make the cost explicit before the choice.",
          },
        ],
        affectedIds: ["scene-card-toy", "setting-pressure-toy"],
      },
    ],
  };
}

File: crates/core/src/model/methodology.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::IsoDateTime;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ConfidenceLevel {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DraftReviewSeverity {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ChangeType {
    Create,
    Update,
    Remove,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SourceReference {
    pub source_id: String,
    pub title: String,
    pub locator: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceLocator {
    pub locator: String,
    pub summary: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceCoverage {
    pub scope: String,
    pub notes: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConfidenceRating {
    pub level: ConfidenceLevel,
    pub rationale: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceCard {
    pub id: String,
    pub kind: EvidenceCardKind,
    pub source: SourceReference,
    pub evidence: Vec<EvidenceLocator>,
    pub observed_technique: String,
    pub source_function: String,
    pub transferable_mechanism: String,
    pub forbidden_copy: Vec<String>,
    pub coverage: EvidenceCoverage,
    pub confidence: ConfidenceRating,
    pub created_at: IsoDateTime,
    pub updated_at: IsoDateTime,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum EvidenceCardKind {
    EvidenceCard,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReaderContract {
    pub id: String,
    pub kind: ReaderContractKind,
    pub promise: String,
    pub progress: String,
    pub payoff: String,
    pub tone: String,
    pub emotional_target: String,
    pub evidence_card_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ReaderContractKind {
    ReaderContract,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct KnowledgeBoundary {
    pub id: String,
    pub kind: KnowledgeBoundaryKind,
    pub pov: String,
    pub narrator_knows: Vec<String>,
    pub reader_knows: Vec<String>,
    pub withheld_information: Vec<String>,
    pub unreliable: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum KnowledgeBoundaryKind {
    KnowledgeBoundary,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SettingPressure {
    pub id: String,
    pub kind: SettingPressureKind,
    pub setting_element: String,
    pub restriction: String,
    pub resource: String,
    pub cost: String,
    pub choice_impact: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SettingPressureKind {
    SettingPressure,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ReferenceMechanism {
    pub id: String,
    pub kind: ReferenceMechanismKind,
    pub evidence_card_id: String,
    pub mechanism: String,
    pub transfer_rules: Vec<String>,
    pub forbidden_copies: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ReferenceMechanismKind {
    ReferenceMechanism,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SceneCard {
    pub id: String,
    pub kind: SceneCardKind,
    pub work_id: String,
    pub title: String,
    pub pov: String,
    pub knowledge_boundary_id: String,
    pub setting_pressure_ids: Vec<String>,
    pub reader_contract_id: String,
    pub character_goals: Vec<String>,
    pub conflict_agenda: String,
    pub revealed_information: Vec<String>,
    pub hidden_information: Vec<String>,
    pub state_change: String,
    pub turn: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SceneCardKind {
    SceneCard,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DraftReviewItem {
    pub id: String,
    pub kind: DraftReviewItemKind,
    pub target_id: String,
    pub severity: DraftReviewSeverity,
    pub category: String,
    pub finding: String,
    pub recommendation: String,
    pub evidence_card_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DraftReviewItemKind {
    DraftReviewItem,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChangeSetEntry {
    pub target_id: String,
    pub change_type: ChangeType,
    pub description: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChangeSet {
    pub id: String,
    pub kind: ChangeSetKind,
    pub created_at: IsoDateTime,
    pub summary: String,
    pub changes: Vec<ChangeSetEntry>,
    pub affected_ids: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ChangeSetKind {
    ChangeSet,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MethodologySeedBundle {
    pub id: String,
    pub kind: MethodologySeedBundleKind,
    pub created_at: IsoDateTime,
    pub updated_at: IsoDateTime,
    pub purpose: String,
    pub evidence_cards: Vec<EvidenceCard>,
    pub reader_contracts: Vec<ReaderContract>,
    pub knowledge_boundaries: Vec<KnowledgeBoundary>,
    pub setting_pressures: Vec<SettingPressure>,
    pub reference_mechanisms: Vec<ReferenceMechanism>,
    pub scene_cards: Vec<SceneCard>,
    pub draft_review_items: Vec<DraftReviewItem>,
    pub change_sets: Vec<ChangeSet>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum MethodologySeedBundleKind {
    MethodologySeedBundle,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EvidenceCardValidationError {
    missing_fields: Vec<String>,
}

impl EvidenceCardValidationError {
    pub fn missing_fields(&self) -> &[String] {
        &self.missing_fields
    }
}

pub fn validate_evidence_card_value(value: &Value) -> Result<(), EvidenceCardValidationError> {
    let mut missing_fields = Vec::new();

    if value.get("evidence").is_none_or(Value::is_null) {
        missing_fields.push("evidence".to_string());
    } else if value
        .get("evidence")
        .and_then(Value::as_array)
        .is_none_or(Vec::is_empty)
    {
        missing_fields.push("evidence".to_string());
    } else if let Some(evidence_items) = value.get("evidence").and_then(Value::as_array) {
        for (index, evidence) in evidence_items.iter().enumerate() {
            if !has_text(evidence.get("locator")) {
                missing_fields.push(evidence_locator_path(index));
            }
            if !has_text(evidence.get("summary")) {
                missing_fields.push(evidence_summary_path(index));
            }
        }
    }

    if value.get("source").is_none_or(Value::is_null) {
        missing_fields.push("source".to_string());
    } else if let Some(source) = value.get("source") {
        if !has_text(source.get("sourceId")) {
            missing_fields.push("source.sourceId".to_string());
        }
        if !has_text(source.get("title")) {
            missing_fields.push("source.title".to_string());
        }
        if !has_text(source.get("locator")) {
            missing_fields.push("source.locator".to_string());
        }
    }

    if value.get("coverage").is_none_or(Value::is_null) {
        missing_fields.push("coverage".to_string());
    } else if let Some(coverage) = value.get("coverage") {
        if !has_text(coverage.get("scope")) {
            missing_fields.push("coverage.scope".to_string());
        }
        if !has_text(coverage.get("notes")) {
            missing_fields.push("coverage.notes".to_string());
        }
    }

    if value.get("confidence").is_none_or(Value::is_null) {
        missing_fields.push("confidence".to_string());
    } else if let Some(confidence) = value.get("confidence") {
        if confidence.get("level").is_none_or(Value::is_null) {
            missing_fields.push("confidence.level".to_string());
        }
        if !has_text(confidence.get("rationale")) {
            missing_fields.push("confidence.rationale".to_string());
        }
    }

    if missing_fields.is_empty() {
        Ok(())
    } else {
        Err(EvidenceCardValidationError { missing_fields })
    }
}

fn has_text(value: Option<&Value>) -> bool {
    value
        .and_then(Value::as_str)
        .is_some_and(|text| !text.is_empty())
}

fn evidence_locator_path(index: usize) -> String {
    format!("evidence[{index}].locator")
}

fn evidence_summary_path(index: usize) -> String {
    format!("evidence[{index}].summary")
}
